contracts/src/v0.8/dev/keeper2_0/KeeperRegistry2_0.sol

```diff
@@ -40,144 +42,172 @@ contract KeeperRegistry2_0 is
    */
   constructor(
     PaymentModel paymentModel,
-    uint256 registryGasOverhead,
     address link,
-    address linkEthFeed,
+    address linkNativeFeed,
     address fastGasFeed,
     address keeperRegistryLogic,
-    Config memory config
-  ) KeeperRegistryBase2_0(paymentModel, registryGasOverhead, link, linkEthFeed, fastGasFeed) {
+    OnChainConfig memory onChainConfig
+  ) KeeperRegistryBase2_0(paymentModel, link, linkNativeFeed, fastGasFeed) {
     KEEPER_REGISTRY_LOGIC = keeperRegistryLogic;
-    setConfig(config);
+    setOnChainConfig(onChainConfig);
   }
 
+  ////////
   // ACTIONS
+  ////////
 
   /**
-   * @notice adds a new upkeep
-   * @param target address to perform upkeep on
-   * @param gasLimit amount of gas to provide the target contract when
-   * performing upkeep
-   * @param admin address to cancel upkeep and withdraw remaining funds
-   * @param checkData data passed to the contract when checking for upkeep
+   * @dev This struct is used to maintain run time information about an upkeep in transmit function
+   * @member upkeep the upkeep struct
+   * @member earlyChecksPassed whether the upkeep passed early checks before perform
+   * @member paymentParams the paymentParams for this upkeep
+   * @member performSuccess whether the perform was successful
+   * @member gasUsed gasUsed by this upkeep in perform
    */
-  function registerUpkeep(
-    address target,
-    uint32 gasLimit,
-    address admin,
-    bytes calldata checkData
-  ) external override returns (uint256 id) {
-    // Executed through logic contract
-    _fallback();
+  struct UpkeepTransmitInfo {
+    Upkeep upkeep;
+    bool earlyChecksPassed;
+    PerformPaymentParams paymentParams;
+    bool performSuccess;
+    uint256 gasUsed;
   }
 
   /**
-   * @notice simulated by keepers via eth_call to see if the upkeep needs to be
-   * performed. If upkeep is needed, the call then simulates performUpkeep
-   * to make sure it succeeds. Finally, it returns the success status along with
-   * payment information and the perform data payload.
-   * @param id identifier of the upkeep to check
-   * @param from the address to simulate performing the upkeep from
+   * @inheritdoc OCR2Abstract
    */
-  function checkUpkeep(uint256 id, address from)
-    external
-    override
-    cannotExecute
-    returns (
-      bytes memory performData,
-      uint256 maxLinkPayment,
-      uint256 gasLimit,
-      uint256 adjustedGasWei,
-      uint256 linkEth
-    )
-  {
-    // Executed through logic contract
-    _fallback();
+  function transmit(
+    bytes32[3] calldata reportContext,
+    bytes calldata report,
+    bytes32[] calldata rs,
+    bytes32[] calldata ss,
+    bytes32 rawVs
+  ) external override whenNotPaused {
+    uint256 gasOverhead = gasleft();
+    HotVars memory hotVars = s_hotVars;
+    if (!s_transmitters[msg.sender].active) revert OnlyActiveTransmitters();
+    _requireExpectedMsgDataLength(report, rs, ss);
```

Comment:
This is a pattern i copied from other OCR contracts. Since we include 16 * msg.data.length in gasOverhead, need it to control a malicious transmitter

Reply:
On second thought we can probably remove this and directly use `report.length` in calculation. Not exact but its a close estimation anyway and extra gas can be covered in `ACCOUNTING_GAS_OVERHEAD`